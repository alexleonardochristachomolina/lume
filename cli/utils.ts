import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
