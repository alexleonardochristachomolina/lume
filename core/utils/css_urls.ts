type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    result += `${code.slice(index, start)}"${newUrl}"`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    const end = code.indexOf(")", start);
    let url = code.slice(start, end).trim();
    // Remove quotes
    if (url.startsWith('"') || url.startsWith("'")) {
      url = url.slice(1, -1);
    }
    return ["url", url, start, end];
  }
}
