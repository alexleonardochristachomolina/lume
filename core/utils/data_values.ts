import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    return searchValue(data, value);
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}

function searchValue(data: Partial<Data>, value: string): unknown {
  if (!value) {
    return;
  }

  if (value.startsWith("=")) {
    let key = value.slice(1);
    [key, value] = parseFallback(key);

    if (!key.includes(".")) {
      return data[key] ?? searchValue(data, value);
    }

    const keys = key.split(".");
    let val = data;
    for (const key of keys) {
      val = val?.[key];
    }
    return val ?? searchValue(data, value);
  }

  if (value.startsWith("$")) {
    let selector = value.slice(1);
    [selector, value] = parseFallback(selector);

    return queryCss(selector, data.page?.document) ?? searchValue(data, value);
  }

  return value;
}

function parseFallback(key: string): [string, string] {
  const fallback = key.indexOf("||");

  if (fallback !== -1) {
    return [
      key.slice(0, fallback).trim(),
      key.slice(fallback + 2).trim(),
    ];
  }

  return [
    key,
    "",
  ];
}

// https://regexr.com/7qnot
const checkForAttrPattern = /^(.+)\s+(?:attr\(([\w\-]+)\))$/;

function queryCss(query: string, document?: Document) {
  const checkResult = query.match(checkForAttrPattern);

  const hasAttr = checkResult?.[0];
  if (hasAttr) {
    const [_, query, name] = checkResult;
    return document?.querySelector(query)?.getAttribute(name);
  }

  return document?.querySelector(query)?.innerHTML;
}
