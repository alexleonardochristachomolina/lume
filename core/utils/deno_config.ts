import { parse } from "../../deps/jsonc.ts";
import { isUrl } from "../utils/path.ts";

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig extends Partial<ImportMap> {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx" | "precompile";
    jsxImportSource?: string;
    types?: string[];
  };
  [key: string]: unknown;
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports,
          scopes: config.scopes,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}
