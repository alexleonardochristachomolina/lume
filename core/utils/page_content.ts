import type { Page } from "../file.ts";

export function insertContent(
  page: Page,
  content: string,
  placeholder?: string,
) {
  const pageContent = page.content;

  if (pageContent && typeof pageContent === "string") {
    if (placeholder && pageContent.includes(placeholder)) {
      page.content = pageContent.replace(
        placeholder,
        content,
      );
    } else {
      page.content += `\n${content}`;
    }
  } else {
    page.content = content;
  }
}
