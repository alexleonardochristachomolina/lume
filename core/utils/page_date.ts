import { getGitDate, parseDate } from "./date.ts";

import type { Page, RawData } from "../file.ts";

/** Returns the Date instance of a file */
export function getPageDate(page: Page): Date {
  const data = page.data as RawData;
  const { date } = data;

  if (date instanceof Date) {
    return date;
  }

  if (typeof date === "number") {
    return new Date(date);
  }

  const { entry } = page.src;
  const info = entry?.getInfo();

  if (typeof date === "string") {
    if (entry && info) {
      switch (date.toLowerCase()) {
        case "git last modified":
          return getGitDate("modified", entry.src) || info.mtime || new Date();
        case "git created":
          // getGitDate("created", ...) uses `git log --diff-filter=A`
          // which returns nothing for files where the latest commit had a merge conflict.
          // Therefore we fallback to last modified time from git history,
          // which is more reliable than file system creation time under most CI environments.
          return getGitDate("created", entry.src) ||
            getGitDate("modified", entry.src) || info.birthtime ||
            new Date();
      }
    }

    try {
      return parseDate(date);
    } catch {
      throw new Error(`Invalid date: ${date} (${entry?.src})`);
    }
  }

  return info?.birthtime || info?.mtime || new Date();
}
