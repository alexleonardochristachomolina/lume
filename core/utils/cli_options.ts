import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
