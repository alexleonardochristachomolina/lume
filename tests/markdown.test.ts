import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import footnotePlugin from "npm:markdown-it-footnote@3.0.3";

Deno.test("Build a markdown site", async (t) => {
  const site = getSite({
    src: "markdown",
    location: new URL("https://example.com/blog"),
  }, {
    markdown: {
      plugins: [footnotePlugin],
      keepDefaultPlugins: true,
      rules: {
        footnote_block_open: () => (
          '<h4 class="mt-3">Footnotes</h4>\n' +
          '<section class="footnotes">\n' +
          '<ol class="footnotes-list">\n'
        ),
      },
    },
  });

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("Build a markdown with hooks", async (t) => {
  const site = getSite({
    src: "markdown",
    location: new URL("https://example.com/blog"),
  });

  site.hooks.addMarkdownItPlugin(footnotePlugin);
  site.hooks.addMarkdownItRule("footnote_block_open", () => (
    '<h4 class="mt-3">Footnotes</h4>\n' +
    '<section class="footnotes">\n' +
    '<ol class="footnotes-list">\n'
  ));

  await build(site);
  await assertSiteSnapshot(t, site);
});
