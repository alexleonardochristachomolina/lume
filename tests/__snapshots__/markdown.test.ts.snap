export const snapshot = {};

snapshot[`Build a markdown site 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/basic.md",
    "/empty.md",
    "/footnote.md",
    "/with-attributes.md",
    "/with-code.md",
    "/with-deflist.md",
    "/with-filter.vto",
    "/with-module.page.js",
    "/with-vento.md",
  ],
}
`;

snapshot[`Build a markdown site 2`] = `[]`;

snapshot[`Build a markdown site 3`] = `
[
  {
    content: "<!DOCTYPE html>
<h1>Normal page</h1>
<ul>
<li>List</li>
<li>Of</li>
<li>Elements</li>
</ul>
",
    data: {
      basename: "basic",
      children: "<h1>Normal page</h1>
<ul>
<li>List</li>
<li>Of</li>
<li>Elements</li>
</ul>
",
      content: "# Normal page

- List
- Of
- Elements
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Basic page",
      url: "/basic/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/basic",
      remote: undefined,
    },
  },
  {
    content: \`<!DOCTYPE html>
<p>Here is a footnote reference,<sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup> and another.<sup class="footnote-ref"><a href="#fn2" id="fnref2">[2]</a></sup></p>
<!-- deno-fmt-ignore -->
<p>Subsequent paragraphs are indented to show that they
belong to the previous footnote.</p>
<h4 class="mt-3">Footnotes</h4>
<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1" class="footnote-item"><p>Here is the footnote. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn2" class="footnote-item"><p>Here's one with multiple blocks. <a href="#fnref2" class="footnote-backref">↩︎</a></p>
</li>
</ol>
</section>
\`,
    data: {
      basename: "footnote",
      children: \`<p>Here is a footnote reference,<sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup> and another.<sup class="footnote-ref"><a href="#fn2" id="fnref2">[2]</a></sup></p>
<!-- deno-fmt-ignore -->
<p>Subsequent paragraphs are indented to show that they
belong to the previous footnote.</p>
<h4 class="mt-3">Footnotes</h4>
<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1" class="footnote-item"><p>Here is the footnote. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn2" class="footnote-item"><p>Here's one with multiple blocks. <a href="#fnref2" class="footnote-backref">↩︎</a></p>
</li>
</ol>
</section>
\`,
      content: "Here is a footnote reference,[^1] and another.[^longnote]

[^1]: Here is the footnote.

[^longnote]: Here's one with multiple blocks.

<!-- deno-fmt-ignore -->
    Subsequent paragraphs are indented to show that they
    belong to the previous footnote.
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/footnote/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/footnote",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<p><a href="#foo" target="_blank">link</a></p>
',
    data: {
      basename: "with-attributes",
      children: '<p><a href="#foo" target="_blank">link</a></p>
',
      content: "[link](#foo){target=_blank}
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/with-attributes/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-attributes",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<pre><code class="language-html">&lt;p&gt;Example&lt;/p&gt;
</code></pre>
<pre><code>&lt;p&gt;Example without color highlight&lt;/p&gt;
</code></pre>
<p>Not indented code</p>
',
    data: {
      basename: "with-code",
      children: '<pre><code class="language-html">&lt;p&gt;Example&lt;/p&gt;
</code></pre>
<pre><code>&lt;p&gt;Example without color highlight&lt;/p&gt;
</code></pre>
<p>Not indented code</p>
',
      content: "\`\`\`html
<p>Example</p>
\`\`\`

\`\`\`
<p>Example without color highlight</p>
\`\`\`

    Not indented code
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/with-code/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-code",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<!-- deno-fmt-ignore-file -->
<dl>
<dt>This is a definition title</dt>
<dd>And this the description</dd>
<dd>Other description</dd>
<dt>Other title</dt>
<dd>And other description</dd>
</dl>
",
    data: {
      basename: "with-deflist",
      children: "<!-- deno-fmt-ignore-file -->
<dl>
<dt>This is a definition title</dt>
<dd>And this the description</dd>
<dd>Other description</dd>
<dt>Other title</dt>
<dd>And other description</dd>
</dl>
",
      content: "<!-- deno-fmt-ignore-file -->
This is a definition title
: And this the description
: Other description

Other title
: And other description
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/with-deflist/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-deflist",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<h1>Module <strong>example</strong></h1>

<div><p>Welcome to this <a href="/">page</a></p></div>
',
    data: {
      basename: "with-filter",
      children: '<h1>Module <strong>example</strong></h1>

<div><p>Welcome to this <a href="/">page</a></p></div>
',
      content: "<h1>{{ title |> md(true) }}</h1>

<div>{{ description |> md }}</div>
",
      date: [],
      description: "Welcome to this [page](/)
",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Module **example**",
      url: "/with-filter/",
    },
    src: {
      asset: false,
      ext: ".vto",
      path: "/with-filter",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<h1>Module example</h1>
<p><a href="/">Back to home</a></p>
',
    data: {
      basename: "with-module",
      children: '<h1>Module example</h1>
<p><a href="/">Back to home</a></p>
',
      content: "default",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      templateEngine: "js,md",
      title: "Module example",
      url: "/with-module/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/with-module",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<!-- deno-fmt-ignore-file -->
<h1>Module <strong>example</strong></h1>
<p>Foo</p>
<ul>
<li>one: <a href="/items/un.html">un</a></li>
<li>two: <a href="/items/dous.html">dous</a></li>
<li>three: <a href="/items/tres.html">tres</a></li>
</ul>
',
    data: {
      basename: "with-vento",
      children: '<!-- deno-fmt-ignore-file -->
<h1>Module <strong>example</strong></h1>
<p>Foo</p>
<ul>
<li>one: <a href="/items/un.html">un</a></li>
<li>two: <a href="/items/dous.html">dous</a></li>
<li>three: <a href="/items/tres.html">tres</a></li>
</ul>
',
      content: "<!-- deno-fmt-ignore-file -->
# {{ title }}

Foo

{{ for title, no of myData }}
- {{ title }}: [{{ no }}](/items/{{ no }}.html)
{{- /for }}
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      myData: [
        "one",
        "two",
        "three",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      templateEngine: "vto,md",
      title: "Module **example**",
      url: "/with-vento/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-vento",
      remote: undefined,
    },
  },
]
`;

snapshot[`Build a markdown with hooks 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/basic.md",
    "/empty.md",
    "/footnote.md",
    "/with-attributes.md",
    "/with-code.md",
    "/with-deflist.md",
    "/with-filter.vto",
    "/with-module.page.js",
    "/with-vento.md",
  ],
}
`;

snapshot[`Build a markdown with hooks 2`] = `[]`;

snapshot[`Build a markdown with hooks 3`] = `
[
  {
    content: "<!DOCTYPE html>
<h1>Normal page</h1>
<ul>
<li>List</li>
<li>Of</li>
<li>Elements</li>
</ul>
",
    data: {
      basename: "basic",
      children: "<h1>Normal page</h1>
<ul>
<li>List</li>
<li>Of</li>
<li>Elements</li>
</ul>
",
      content: "# Normal page

- List
- Of
- Elements
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Basic page",
      url: "/basic/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/basic",
      remote: undefined,
    },
  },
  {
    content: \`<!DOCTYPE html>
<p>Here is a footnote reference,<sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup> and another.<sup class="footnote-ref"><a href="#fn2" id="fnref2">[2]</a></sup></p>
<!-- deno-fmt-ignore -->
<p>Subsequent paragraphs are indented to show that they
belong to the previous footnote.</p>
<h4 class="mt-3">Footnotes</h4>
<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1" class="footnote-item"><p>Here is the footnote. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn2" class="footnote-item"><p>Here's one with multiple blocks. <a href="#fnref2" class="footnote-backref">↩︎</a></p>
</li>
</ol>
</section>
\`,
    data: {
      basename: "footnote",
      children: \`<p>Here is a footnote reference,<sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup> and another.<sup class="footnote-ref"><a href="#fn2" id="fnref2">[2]</a></sup></p>
<!-- deno-fmt-ignore -->
<p>Subsequent paragraphs are indented to show that they
belong to the previous footnote.</p>
<h4 class="mt-3">Footnotes</h4>
<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1" class="footnote-item"><p>Here is the footnote. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
</li>
<li id="fn2" class="footnote-item"><p>Here's one with multiple blocks. <a href="#fnref2" class="footnote-backref">↩︎</a></p>
</li>
</ol>
</section>
\`,
      content: "Here is a footnote reference,[^1] and another.[^longnote]

[^1]: Here is the footnote.

[^longnote]: Here's one with multiple blocks.

<!-- deno-fmt-ignore -->
    Subsequent paragraphs are indented to show that they
    belong to the previous footnote.
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/footnote/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/footnote",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<p><a href="#foo" target="_blank">link</a></p>
',
    data: {
      basename: "with-attributes",
      children: '<p><a href="#foo" target="_blank">link</a></p>
',
      content: "[link](#foo){target=_blank}
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/with-attributes/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-attributes",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<pre><code class="language-html">&lt;p&gt;Example&lt;/p&gt;
</code></pre>
<pre><code>&lt;p&gt;Example without color highlight&lt;/p&gt;
</code></pre>
<p>Not indented code</p>
',
    data: {
      basename: "with-code",
      children: '<pre><code class="language-html">&lt;p&gt;Example&lt;/p&gt;
</code></pre>
<pre><code>&lt;p&gt;Example without color highlight&lt;/p&gt;
</code></pre>
<p>Not indented code</p>
',
      content: "\`\`\`html
<p>Example</p>
\`\`\`

\`\`\`
<p>Example without color highlight</p>
\`\`\`

    Not indented code
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/with-code/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-code",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<!-- deno-fmt-ignore-file -->
<dl>
<dt>This is a definition title</dt>
<dd>And this the description</dd>
<dd>Other description</dd>
<dt>Other title</dt>
<dd>And other description</dd>
</dl>
",
    data: {
      basename: "with-deflist",
      children: "<!-- deno-fmt-ignore-file -->
<dl>
<dt>This is a definition title</dt>
<dd>And this the description</dd>
<dd>Other description</dd>
<dt>Other title</dt>
<dd>And other description</dd>
</dl>
",
      content: "<!-- deno-fmt-ignore-file -->
This is a definition title
: And this the description
: Other description

Other title
: And other description
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/with-deflist/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-deflist",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<h1>Module <strong>example</strong></h1>

<div><p>Welcome to this <a href="/">page</a></p></div>
',
    data: {
      basename: "with-filter",
      children: '<h1>Module <strong>example</strong></h1>

<div><p>Welcome to this <a href="/">page</a></p></div>
',
      content: "<h1>{{ title |> md(true) }}</h1>

<div>{{ description |> md }}</div>
",
      date: [],
      description: "Welcome to this [page](/)
",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Module **example**",
      url: "/with-filter/",
    },
    src: {
      asset: false,
      ext: ".vto",
      path: "/with-filter",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<h1>Module example</h1>
<p><a href="/">Back to home</a></p>
',
    data: {
      basename: "with-module",
      children: '<h1>Module example</h1>
<p><a href="/">Back to home</a></p>
',
      content: "default",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      templateEngine: "js,md",
      title: "Module example",
      url: "/with-module/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/with-module",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<!-- deno-fmt-ignore-file -->
<h1>Module <strong>example</strong></h1>
<p>Foo</p>
<ul>
<li>one: <a href="/items/un.html">un</a></li>
<li>two: <a href="/items/dous.html">dous</a></li>
<li>three: <a href="/items/tres.html">tres</a></li>
</ul>
',
    data: {
      basename: "with-vento",
      children: '<!-- deno-fmt-ignore-file -->
<h1>Module <strong>example</strong></h1>
<p>Foo</p>
<ul>
<li>one: <a href="/items/un.html">un</a></li>
<li>two: <a href="/items/dous.html">dous</a></li>
<li>three: <a href="/items/tres.html">tres</a></li>
</ul>
',
      content: "<!-- deno-fmt-ignore-file -->
# {{ title }}

Foo

{{ for title, no of myData }}
- {{ title }}: [{{ no }}](/items/{{ no }}.html)
{{- /for }}
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      myData: [
        "one",
        "two",
        "three",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      templateEngine: "vto,md",
      title: "Module **example**",
      url: "/with-vento/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/with-vento",
      remote: undefined,
    },
  },
]
`;
